src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
   }
 
   override def concat[V1 >: V](that: scala.IterableOnce[(K, V1)]): HashMap[K, V1] = {
-    // TODO PERF We could avoid recomputing entry hash's when `that` is another `HashMap`
-    super.concat(that)
+
+    /** Fall back to slow concatenation in case of error, or non-hashmap `that` */
+    def slowConcat = {
+      val builder = mapFactory.newBuilder[K, V1]
+      builder ++= this
+      builder ++= that
+      builder.result()
+    }
+
+    that match {
+      case hm: HashMap[K, V1] =>
+        var newHash = cachedJavaKeySetHashCode + hm.cachedJavaKeySetHashCode
+
+        /** Recursively, immutably concatenates two MapNodes, node-by-node as to visit the least number of nodes possible */
+        def concat(left: MapNode[K, V1], right: MapNode[K, V1], shift: Int): MapNode[K, V1] =
```

Comment:
We might want to make this constant stack to be consistent with the other trie traversals.

Do you think we could extract some analog of `ChampBaseIterator` to capture the parts of traversal that are indepdendant of the `Map`/`Set` distinction, and implement this in terms of that?

Reply:
Yeah I've been thinking a lot about that. That is probably the right way to go so that these classes don't have a dozen slightly different intricate bit fiddling implementations, I just have a hard time thinking how it would look